File: python/yacht/yacht.py
# Score categories
# Change the values as you see fit
YACHT = 'YACHT'
ONES = 'ONES'
TWOS = 'TWOS'
THREES = 'THREES'
FOURS = 'FOURS'
FIVES = 'FIVES'
SIXES = 'SIXES'
FULL_HOUSE = 'FULL'
FOUR_OF_A_KIND = 'FOUR'
LITTLE_STRAIGHT = 'LITTLE'
BIG_STRAIGHT = 'BIG'
CHOICE = 'CHOICE'


def score(dice, category):
    is_f = TESTING_FUNCTIONS[category]
    scoring_f = SCORING_FUNCTIONS[category]
    if not is_f(dice):
        return 0
    return scoring_f(dice)


# Scoring functions

def score_number(dice, number):
    d = to_dict(dice)
    n = d.get(number, 0)
    return n * number


def score_full_house(dice):
    return sum(dice)


def score_four_of_a_kind(dice):
    d = to_dict(dice)
    number = [die for die, n in d.items() if n >= 4][0]
    return number * 4  # always counts as four


def score_little_straight(dice):
    return 30


def score_big_straight(dice):
    return 30


def score_choice(dice):
    return sum(dice)


def score_yacht(dice):
    return 50


# Testing functions

def is_number(dice, number):
    return True


def is_full_house(dice):
    d = to_dict(dice)
    if len(d) != 2:
        return False
    return sorted(d.values()) == [2, 3]


def is_four_of_a_kind(dice):
    d = to_dict(dice)
    most = max(d.values())
    return most >= 4


def is_little_straight(dice):
    return sorted(dice) == [1, 2, 3, 4, 5]


def is_big_straight(dice):
    return sorted(dice) == [2, 3, 4, 5, 6]


def is_choice(dice):
    return True


def is_yacht(dice):
    d = to_dict(dice)
    most = max(d.values())
    return most >= 5


# Category dictionaries

TESTING_FUNCTIONS = {
    YACHT: is_yacht,
    ONES: lambda dice: is_number(dice, 1),
    TWOS: lambda dice: is_number(dice, 2),
    THREES: lambda dice: is_number(dice, 3),
    FOURS: lambda dice: is_number(dice, 4),
    FIVES: lambda dice: is_number(dice, 5),
    SIXES: lambda dice: is_number(dice, 6),
    FULL_HOUSE: is_full_house,
    FOUR_OF_A_KIND: is_four_of_a_kind,
    LITTLE_STRAIGHT: is_little_straight,
    BIG_STRAIGHT: is_big_straight,
    CHOICE: is_choice,
}


SCORING_FUNCTIONS = {
    YACHT: score_yacht,
    ONES: lambda dice: score_number(dice, 1),
    TWOS: lambda dice: score_number(dice, 2),
    THREES: lambda dice: score_number(dice, 3),
    FOURS: lambda dice: score_number(dice, 4),
    FIVES: lambda dice: score_number(dice, 5),
    SIXES: lambda dice: score_number(dice, 6),
    FULL_HOUSE: score_full_house,
    FOUR_OF_A_KIND: score_four_of_a_kind,
    LITTLE_STRAIGHT: score_little_straight,
    BIG_STRAIGHT: score_big_straight,
    CHOICE: score_choice,
}


# Helper functions

def to_dict(dice):
    d = {}
    for die in dice:
        d[die] = d.get(die, 0) + 1
    return d

File: haskell/yacht/src/Yacht.hs
module Yacht (yacht, Category(..)) where

import Data.Function ((&))
import Data.Functor ((<&>))
import Data.List (sort)
import Data.Maybe (listToMaybe, fromMaybe)
import Control.Arrow ((>>>))
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map

type Die = Int
type Dice = [Die]
type Score = Int
type Scorer = Dice -> Score

data Category = Ones
              | Twos
              | Threes
              | Fours
              | Fives
              | Sixes
              | FullHouse
              | FourOfAKind
              | LittleStraight
              | BigStraight
              | Choice
              | Yacht

yacht :: Category -> Dice -> Score
yacht category =
  case category of
    Ones -> scoreNum 1
    Twos -> scoreNum 2
    Threes -> scoreNum 3
    Fours -> scoreNum 4
    Fives -> scoreNum 5
    Sixes -> scoreNum 6
    FullHouse -> scoreFullHouse
    FourOfAKind -> scoreNOfAKind 4
    LittleStraight -> scoreLittleStraight
    BigStraight -> scoreBigStraight
    Choice -> scoreChoice
    Yacht -> scoreYacht

scoreNum :: Die -> Scorer
scoreNum num = filter (== num) >>> sum

scoreFullHouse :: Scorer
scoreFullHouse dice =
  let diceCounts = counts dice & Map.elems & sort
  in if diceCounts == [2, 3] then sum dice else 0
     
scoreLittleStraight, scoreBigStraight :: Scorer
scoreLittleStraight = scoreMatchingDice [1,2,3,4,5] littleStrightScore
scoreBigStraight = scoreMatchingDice [2,3,4,5,6] bigStraightScore

scoreChoice :: Scorer
scoreChoice dice = sum dice

scoreYacht :: Scorer
scoreYacht dice =
  let firstMaybe = listToMaybe dice
      allMatchMaybe = firstMaybe <&> flip allMatch dice
  in case allMatchMaybe of
    Just True -> yachtScore
    _ -> 0

littleStrightScore, bigStraightScore, yachtScore :: Score
littleStrightScore = 30
bigStraightScore = 30
yachtScore = 50

scoreMatchingDice :: Dice -> Score -> Scorer
scoreMatchingDice matchDice score dice =
  if sort dice == matchDice then score else 0

scoreNOfAKind :: Int -> Scorer
scoreNOfAKind n dice =
  let diceWithCounts = counts dice & Map.assocs
      countIsAtLeastN (_, c) = c >= n
      dieWithAtLeastN = diceWithCounts & filter countIsAtLeastN & map fst & listToMaybe
      scoreMaybe = dieWithAtLeastN <&> (* n)
  in fromMaybe 0 scoreMaybe

counts :: Ord a => [a] -> Map a Int
counts =
  let insert die set = Map.insertWith (+) die 1 set
  in foldr insert Map.empty

allMatch :: Eq a => a -> [a] -> Bool
allMatch a = map (== a) >>> and
